flake8_trio/visitors/visitor91x.py

```diff
@@ -54,148 +106,232 @@ def __init__(self, *args: Any, **kwargs: Any):
+
+    @body_guaranteed_once.setter
+    def body_guaranteed_once(self, value: bool):
+        self.loop_state.body_guaranteed_once = value
+
+    @property
+    def has_break(self) -> bool:
+        return self.loop_state.has_break
+
+    @has_break.setter
+    def has_break(self, value: bool):
+        self.loop_state.has_break = value
+
+    @property
+    def uncheckpointed_before_continue(self) -> set[Statement]:
+        return self.loop_state.uncheckpointed_before_continue
+
+    @uncheckpointed_before_continue.setter
+    def uncheckpointed_before_continue(self, value: set[Statement]):
+        self.loop_state.uncheckpointed_before_continue = value
 
-    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+    @property
+    def uncheckpointed_before_break(self) -> set[Statement]:
+        return self.loop_state.uncheckpointed_before_break
+
+    @uncheckpointed_before_break.setter
+    def uncheckpointed_before_break(self, value: set[Statement]):
+        self.loop_state.uncheckpointed_before_break = value
+
+    @property
+    def artificial_errors(self) -> set[cst.Return | cst.FunctionDef | cst.Yield]:
+        return self.loop_state.artificial_errors
+
+    @artificial_errors.setter
+    def artificial_errors(self, value: set[cst.Return | cst.FunctionDef | cst.Yield]):
+        self.loop_state.artificial_errors = value  # pragma: no cover
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> None:
         # don't lint functions whose bodies solely consist of pass or ellipsis
-        if has_decorator(node, "overload", "fixture") or empty_body(node.body):
+        if func_has_decorator(node, "overload", "fixture") or func_empty_body(node):
             return
 
-        self.save_state(node)
-        self.set_state(self.default, copy=True)
+        self.save_state(
+            node,
+            "has_yield",
+            "safe_decorator",
+            "async_function",
+            "uncheckpointed_statements",
+            "loop_state",
+            "try_state",
+            copy=True,
+        )
+        self.uncheckpointed_statements = set()
+        self.has_yield = self.safe_decorator = False
+        self.loop_state = LoopState()
 
-        # disable checks in asynccontextmanagers by saying the function isn't async
-        self.async_function = not fnmatch_qualified_name(
-            node.decorator_list, *self.options.no_checkpoint_warning_decorators
+        self.async_function = (
+            node.asynchronous is not None
+            and not fnmatch_qualified_name_cst(
+                node.decorators, *self.options.no_checkpoint_warning_decorators
+            )
         )
         if not self.async_function:
             return
 
+        pos = self.get_metadata(PositionProvider, node).start
         self.uncheckpointed_statements = {
-            Statement("function definition", node.lineno, node.col_offset)
+            Statement("function definition", pos.line, pos.column)
```

Comment:
will quite possibly change this in future PR's to save nodes instead of statements, just gotta figure out how that plays with tests.